### ISSUE TITLE
SWITCH freezes latent blocks during RL; gradients do not traverse the recurrence

### ISSUE THREAD

**Issue by microprediction** [microprediction/latentreasoning#67]
## Reviewed source

Default-branch commit: [`11b700b`](https://github.com/microprediction/latentreasoning/commit/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51)

Affected text:

- [`training.html` lines 367–378](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/training.html#L367-L378)
- Related framing at [`training.html` lines 157–180](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/training.html#L157-L180)
- [`bibliography.html` lines 1108–1112](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/bibliography.html#L1108-L1112)

The current SWITCH summary correctly says that latent positions have no policy-gradient term. It then says:

> “Gradients still flow through them from the tokens on either side, the pathwise route from the previous section, confined to the inside of a block.”

That is not what the reported Switch-GRPO implementation does. The latent state affects downstream text generation, but the gradient pass deliberately detaches the latent computation.

I searched open and closed issues for SWITCH, Switch-GRPO, frozen KV, latent positions, and `no_grad`; I found no issue covering this implementation distinction.

## The paper explicitly freezes latent segments during RL

In [Yang et al., arXiv:2606.13106v1](https://arxiv.org/html/2606.13106), the likelihood factorization is over text positions only: `<swi>`, `</swi>`, and visible answer tokens have categorical policy ratios, while latent positions have none ([§3.3](https://arxiv.org/html/2606.13106#S3.SS3)).

More importantly, the implementation section states that each rollout is split at switch boundaries, then:

- latent segments run inside `torch.no_grad()` and store no autograd graph;
- text segments run with gradients enabled;
- a streaming KV cache is passed between segments.

See [Appendix A, memory-segmented backward pass](https://arxiv.org/html/2606.13106#A1), especially the paragraph beginning “Storing one autograd graph...”. The paper's own limitations section is unambiguous:

> “Switch-GRPO’s gradient flows through the text segments of each rollout only; latent positions contribute via a frozen KV cache ... so the latent representation itself is shaped primarily by the Phase 2 curriculum rather than directly by the RL objective.”

See [§6, Limitations](https://arxiv.org/html/2606.13106#S6).

## Conditioning on a latent state is not differentiating through its construction

For one latent transition and a later text token, write

```text
h = F_θ(h₀),
ℓ(θ) = log π_θ(y | h).
```

Full pathwise differentiation through the recurrence contains two terms:

```text
dℓ/dθ = ∂ℓ/∂θ + (∂h/∂θ)ᵀ ∂ℓ/∂h.
```

Replaying `h` or its KV cache as a frozen rollout prefix retains only the first term. The text-token loss is evaluated *conditional on* the realized latent state, so the state affects the logits and reward trajectory, but the Jacobian `∂h/∂θ` is absent. That is precisely what `torch.no_grad()` accomplishes.

The distinction is not semantic. The site's claimed “pathwise route ... inside a block” is the second term above; Switch's memory-segmented implementation omits it.

Shared weights mean a text-segment update can still change future latent rollouts when the next batch is generated. RL can also learn whether to open a block and can improve visible-token predictions conditioned on frozen latent states. Those are indirect effects of parameter sharing and new on-policy rollouts, not backpropagation through the latent recurrence in the current update.

## The primary source is internally inconsistent, so the implementation detail matters

The paper's abstract says the objective “propagates gradients through recurrent latent computation,” and some main-text prose uses similar language. But its algorithmic implementation and limitations explicitly say the opposite: latent segments run under `no_grad`, the KV cache is frozen, and the representation is not directly shaped by the RL objective.

The site currently resolves that inconsistency in favor of the abstract rather than the actual backward pass. A claim-level review should prefer the specified computation.

## Consequence

SWITCH is evidence that on-policy RL can improve a model whose *rollouts contain* hidden-state recurrence and whose discrete boundary decisions have tractable ratios. It is not evidence that Switch-GRPO assigns a pathwise terminal-reward gradient through the recurrent latent steps.

This also narrows the interpretation of the reported 66.7% → 79.3% MATH-500 gain. Because the study does not ablate full BPTT against detached latent replay, that gain cannot identify direct optimization of the latent transition itself. It may come from switch selection, visible-token policy improvement, shared-weight spillover into later rollouts, or their combination. The paper itself says the latent representation remains shaped primarily by curriculum SFT.

## Concrete correction

For example:

> SWITCH defines GRPO ratios only at ordinary text positions, including the entry and exit tokens. In the reported memory-segmented implementation, latent blocks run under `torch.no_grad()` and their KV caches are replayed as frozen conditioning for the surrounding text segments. Latent states therefore influence the rollout and downstream token likelihoods, but the RL backward pass does not traverse the latent recurrence. Parameter sharing and subsequent on-policy rollouts can still alter latent behavior indirectly. The paper's abstract describes this as propagating gradients through recurrent computation, but its implementation and limitations explicitly qualify that claim.

The bibliography entry can remain concise, but the training page should remove the claimed pathwise route and preserve this implementation limitation.

**Comment by microprediction** [comment]
Correct, and the strongest of this batch. Fixed in 1185867.

Verified against arXiv:2606.13106. Appendix A: "Latent segments run inside torch.no_grad() and store nothing; text segments run with gradient enabled... the only price we pay is that the latent positions contribute to the gradient pass only through the KV cache they hand off to the next text segment." And the limitations section: "latent positions contribute via a frozen KV cache, so the latent representation itself is shaped primarily by the Phase 2 curriculum rather than directly by the RL objective."

The page said gradients still flow through the latent positions by the pathwise route. That names exactly the term no_grad removes. You are also right that the paper contradicts itself, since the abstract says the objective propagates gradients through recurrent latent computation, and the page had resolved that in favour of the abstract. It now reports both and says which one the implementation follows.
